File: SwitchBasemapSample/objective-c/SwitchBasemapSample/Controllers/BasemapsCollectionViewController.m
// Copyright 2014 ESRI
//
// All rights reserved under the copyright laws of the United States
// and applicable international laws, treaties, and conventions.
//
// You may freely redistribute and use this sample code, with or
// without modification, provided you include the original copyright
// notice and use restrictions.
//
// See the use restrictions at http://help.arcgis.com/en/sdk/10.0/usageRestrictions.htm
//

#import "BasemapsCollectionViewController.h"
#import "AppConstants.h"
#import "SVProgressHUD.h"


@interface BasemapsCollectionViewController ()

@property (nonatomic, strong) NSArray *basemaps;
@property (nonatomic, strong) AGSPortal *portal;
@property (nonatomic, strong) PortalBasemapHelper *portalBasemapHelper;
@property (weak, nonatomic) IBOutlet UICollectionView *collectionView;
@property (nonatomic, strong) AGSCredential *credential;

@end

@implementation BasemapsCollectionViewController

//in order to keep a single instance of the view controller
//creating a sharedInstance variable and will assign it the
//view controller once created
static id sharedInstance = nil;

//once the view controller is instantiated through the storyboard
//the following method is called for initialization
- (id)initWithCoder:(NSCoder *)coder
{
    self = [super initWithCoder:coder];
    if (self) {
        //storing the instance created as the sharedInstance inside a dispatch_once
        //block, in order to assign only the first time
        static dispatch_once_t onceToken;
        dispatch_once(&onceToken, ^{
            sharedInstance = self;
        });
    }
    return self;
}

//the following method is called everytime an object is instantiated using the initwithcoder method
//and we are using this method to return the sharedInstance (if exists) instead of the newly instantiated object
-(id)awakeAfterUsingCoder:(NSCoder *)aDecoder {
    if (sharedInstance) {
        return sharedInstance;
    }
    return self;
}

// in iOS7 this gets called and hides the status bar so the view does not go under the top iPhone status bar
- (BOOL)prefersStatusBarHidden
{
    return YES;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    // Do any additional setup after loading the view.
}

-(void)viewDidAppear:(BOOL)animated {
    //viewDidLoad called everytime you trigger the segue
    //so in order to avoid reloading data everytime add a condition
    if (!self.basemaps) {
        [self loadBasemaps];
    }
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)loadBasemaps {
    //show the progress hud
    [SVProgressHUD showWithStatus:@"Loading"];
    
    //instantiate the portalBasemapHelper if nil
    if (!self.portalBasemapHelper) {
        self.portalBasemapHelper = [[PortalBasemapHelper alloc] init];
    }
    
    //prepare for connection to the portal
    NSURL *portalUrl = [NSURL URLWithString:kPortalUrl];
    //if required create a credential
    //self.credential = [[AGSCredential alloc] initWithUser:kUserName password:kPassword];
    self.credential = nil;
    __weak __typeof(self) weakSelf = self;
    [self.portalBasemapHelper fetchBasemapsFromPortal:portalUrl withCredential:self.credential completion:^(NSArray<AGSBasemap *> *basemaps, NSError *error) {
        [SVProgressHUD dismiss];
        if (error) {
            [[[UIAlertView alloc] initWithTitle:@"Error" message:error.localizedDescription delegate:nil cancelButtonTitle:@"Ok" otherButtonTitles: nil] show];
        }
        else {
            //populate the basemaps array and reload data
            weakSelf.basemaps = basemaps;
            [weakSelf.collectionView reloadData];
        }
    }];
}

#pragma mark - dataSource methods

-(NSInteger)numberOfSectionsInCollectionView:(UICollectionView *)collectionView {
    return 1;
}

-(NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    //an extra item cell in case there are more results
    return self.basemaps.count;
}

-(UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    
    //item cell for the regular data
    NSString *reusableString = @"Cell";
    UICollectionViewCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:reusableString forIndexPath:indexPath];
    
    AGSBasemap *basemap = [self.basemaps objectAtIndex:indexPath.item];
    cell.tag = indexPath.row + 1000;  // add 1000 to prevent conflicts with other tag operations
    
    UIImageView *imageView = (UIImageView*)[cell viewWithTag:1];
    [imageView setImage:basemap.item.thumbnail.image];
    if (!basemap.item.thumbnail.image) {
        //if we don't have an image yet, load it
        [basemap.item.thumbnail loadWithCompletion:^(NSError * _Nullable error) {
            //make sure this is loaded and we're still on the correct row
            if (!error && cell.tag == indexPath.row + 1000) {
                [imageView setImage:basemap.item.thumbnail.image];
            }
        }];
    }
    
    UILabel *label = (UILabel*)[cell viewWithTag:2];
    [label setText:basemap.item.title];
    
    //add shadow to the cell
    [cell.layer setShadowOpacity:1];
    [cell.layer setShadowColor:[UIColor lightGrayColor].CGColor];
    [cell.layer setShadowRadius:2];
    [cell.layer setShadowOffset:CGSizeMake(2, 2)];
    [cell setClipsToBounds:NO];
    
    return cell;
}

#pragma mark - delegate methods

-(void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath {
    AGSBasemap *basemap = (AGSBasemap*)[self.basemaps objectAtIndex:indexPath.item];
    if (basemap) {
        [self.delegate basemapPickerController:self didSelectBasemap:basemap];
    }
}

-(CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout sizeForItemAtIndexPath:(NSIndexPath *)indexPath {
    //for iPad showing 3 item cells each row
    if ([self isIPad]) {
        double width = (self.collectionView.frame.size.width - 40)/3.0;
        return CGSizeMake(width, width);
    }
    //for iPhone showing 2 item cells each row for landscape
    //and 3 item cells for portrait
    else {
        double width;
        if ([[UIDevice currentDevice] orientation] == UIInterfaceOrientationLandscapeLeft || [[UIDevice currentDevice] orientation] == UIInterfaceOrientationLandscapeRight) {
            width = (self.collectionView.frame.size.width - 40)/3.0;
        }
        else {
            width = (self.collectionView.frame.size.width - 30)/2.0;
        }
        return CGSizeMake(width, width);
    }
}

#pragma mark - actions

-(IBAction)cancel:(id)sender {
    [self.delegate basemapPickerControllerDidCancel:self];
}

//reload the data to adjust the layout based on orientation
-(void)didRotateFromInterfaceOrientation:(UIInterfaceOrientation)fromInterfaceOrientation {
    [self.collectionView reloadData];
}

#pragma mark - internal

-(BOOL)isIPad {
    return ([UIDevice currentDevice].userInterfaceIdiom == UIUserInterfaceIdiomPad);
}

@end

File: arcgis-ios-sdk-samples/Cloud & Portal/Org basemap picker/BasemapsCollectionViewController.swift
//
// Copyright 2016 Esri.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import UIKit
import ArcGIS

protocol BasemapsCollectionVCDelegate: class {
    
    func basemapsCollectionViewController(basemapsCollectionViewController: BasemapsCollectionViewController, didSelectBasemap basemap: AGSBasemap)
    func basemapsCollectionViewControllerDidCancel(basemapsCollectionViewController: BasemapsCollectionViewController)
}

class BasemapsCollectionViewController: UIViewController, UICollectionViewDataSource, UICollectionViewDelegate {
    
    @IBOutlet var collectionView: UICollectionView!
    
    var portalURLString: String!
    var anonymousUser: Bool = true
    
    private var portal: AGSPortal!
    private var basemapsGroup: AGSPortalGroup!
    private var basemaps = [AGSPortalItem]()
    private var resultSet: AGSPortalQueryResultSet!
    
    weak var delegate: BasemapsCollectionVCDelegate?
    
    override func prefersStatusBarHidden() -> Bool {
        return true
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.loadPortal()
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    private func loadPortal() {
        self.portal = AGSPortal(URL: NSURL(string: self.portalURLString)!, loginRequired: !self.anonymousUser)
        self.portal.loadWithCompletion { [weak self] (error: NSError?) in
            if let error = error {
                print(error)
            }
            else {
                self?.fetchBasemapsGroup()
            }
        }
    }
    
    private func fetchBasemapsGroup() {
        if let queryString = self.portal.portalInfo?.basemapGalleryGroupQuery {
            let queryParameters = AGSPortalQueryParameters(query: queryString)
            self.portal.findGroupsWithQueryParameters(queryParameters, completion: { (resultSet: AGSPortalQueryResultSet?, error: NSError?) in
                if let error = error {
                    print(error)
                }
                else {
                    if let basemapsGroup = resultSet?.results?[0] as? AGSPortalGroup {
                        self.basemapsGroup = basemapsGroup
                        let queryParameters = AGSPortalQueryParameters(forItemsInGroup: "\(self.basemapsGroup.groupID!)")
                        self.fetchBasemaps(queryParameters)
                    }
                }
            })
        }
    }
    
    private func fetchBasemaps(queryParameters: AGSPortalQueryParameters) {
        queryParameters.limit = 5
        self.portal.findItemsWithQueryParameters(queryParameters) { (resultSet: AGSPortalQueryResultSet?, error: NSError?) in
            if let error = error {
                print(error)
            }
            else {
                if let basemaps = resultSet?.results as? [AGSPortalItem] {
                    self.loadBasemaps(basemaps)
                    self.resultSet = resultSet!
                }
            }
        }
    }
    
    private func loadBasemaps(basemaps: [AGSPortalItem]) {
        AGSLoadObjects(basemaps) { (succeed: Bool) in
            //reload collection view
            self.basemaps.appendContentsOf(basemaps)
            self.collectionView.reloadData()
        }
    }
    
    private func loadMoreBasemaps() {
        if let queryParameters = self.resultSet.nextQueryParameters {
            self.fetchBasemaps(queryParameters)
        }
    }
    
    //MARK: - UICollectionViewDataSource
    
    func collectionView(collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        if let _ = self.resultSet?.nextQueryParameters {
            return self.basemaps.count + 1
        }
        else {
            return self.basemaps.count
        }
    }
    
    func collectionView(collectionView: UICollectionView, cellForItemAtIndexPath indexPath: NSIndexPath) -> UICollectionViewCell {
        
        if indexPath.row == self.basemaps.count {
            let cell = collectionView.dequeueReusableCellWithReuseIdentifier("PlusCell", forIndexPath: indexPath)
            
            cell.layer.borderColor = UIColor.grayColor().CGColor
            cell.layer.borderWidth = 1
            
            return cell
        }
        else {
            let cell = collectionView.dequeueReusableCellWithReuseIdentifier("BasemapCell", forIndexPath: indexPath) as! BasemapCell
            
            let portalItem = self.basemaps[indexPath.item]
            
            cell.label.text = portalItem.title
            
            //image
            if let image = portalItem.thumbnail?.image {
                cell.imageView.image = image
            }
            else {
                cell.imageView.image = nil
                portalItem.thumbnail?.loadWithCompletion({ (error: NSError?) in
                    if let error = error {
                        print(error)
                    }
                    else {
                        if let cell = collectionView.cellForItemAtIndexPath(indexPath) as? BasemapCell {
                            cell.imageView.image = portalItem.thumbnail?.image
                        }
                    }
                })
            }
            
            cell.layer.borderColor = UIColor.grayColor().CGColor
            cell.layer.borderWidth = 1
            
            return cell
        }
    }
    
    //MARK: - UICollectionViewDelegate
    
    func collectionView(collectionView: UICollectionView, didSelectItemAtIndexPath indexPath: NSIndexPath) {
        
        if indexPath.row == self.basemaps.count {
            self.loadMoreBasemaps()
        }
        else {
            let basemap = AGSBasemap(item: self.basemaps[indexPath.row])
            self.delegate?.basemapsCollectionViewController(self, didSelectBasemap: basemap)
        }
    }
    
    //MARK: - Actions
    
    @IBAction private func cancelAction() {
        self.delegate?.basemapsCollectionViewControllerDidCancel(self)
    }
    
    //
    //Cell resize logic for device rotation
    //
    
    //MARK: - UICollectionViewDelegateFlowLayout
    
    func collectionView(collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAtIndexPath indexPath: NSIndexPath) -> CGSize {
        
        let collectionViewSize = collectionView.bounds.size
        let defaultCellSize:CGFloat = 140
        let defaultCellSpacing:CGFloat = 10
        let cellSizePlusSpacing = defaultCellSize + defaultCellSpacing
        
        //find out number of possible items in one row
        let n = floor((collectionViewSize.width - defaultCellSpacing) / (cellSizePlusSpacing))
        
        //extra space per cell
        let extra = (collectionViewSize.width - (n * cellSizePlusSpacing) - defaultCellSpacing) / n
        
        return CGSize(width: defaultCellSize + extra, height: defaultCellSize + extra)
    }
    
    //MARK: -
    
    override func viewWillTransitionToSize(size: CGSize, withTransitionCoordinator coordinator: UIViewControllerTransitionCoordinator) {
        
        coordinator.animateAlongsideTransition({ (_: UIViewControllerTransitionCoordinatorContext) in
            self.collectionView.performBatchUpdates(nil, completion: nil)
            }, completion: nil)
    }
}
